### PR TITLE
[TECH DEBT] Refactor errors

### PR DESCRIPTION
**Pull request by rob-maron**
### This PR: 

Is the first step towards getting better logging/errors.

- Switches everything from `snafu` to `thiserror`. This should get rid of the errors akin to "Libp2p-specific errors" and "Error in the Kademlia DHT"
- Deduplicates some warnings/errors. Some errors will log and then return, where they got logged a second time. I plan to do more of this in a coming PR, along with adding more context where necessary
- Removes unused errors and error variants

### This PR does not: 

- Refine log levels
- Change things to use `anyhow` anywhere
- Apart from formatting, does not touch existing, in-use log messages (besides adding context) or log types. 

### Key places to review: 

The only really important errors are the HotShot ones, so everything mentioning `HotShotError`

### How to test this PR: 

I've run this in the sequencer. I didn't see any unused error variants or anything

### CODE REVIEW

File: crates/types/src/data.rs [hunk 1]
```diff
@@ -88,6 +88,12 @@ impl ConsensusTime for ViewNumber {
     }
 }
 
+impl Display for ViewNumber {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
```
Comment (ss-es) [inline crates/types/src/data.rs:96]: not inherently against this, but is there a reason not to `#[derive(Display)]`?

**Review by ss-es — APPROVED** [review]
no opinion on `thiserror` vs `snafu` but the PR itself looks good! just one comment

File: crates/types/src/data.rs (hunk 1 above)
Comment (rob-maron) [reply crates/types/src/data.rs:96]: Oh, it doesn't have a `Display` derive

Comment (ss-es) [reply crates/types/src/data.rs:96]: oh oops, my bad -- yeah you're right, nevermind: https://doc.rust-lang.org/std/fmt/trait.Display.html

**Pull request merged**